### PR TITLE
579 better error when private org

### LINKED ISSUE #579: Better error message when org is private and user does not have access
It should also allow user to logout and try a different account

### PR DESCRIPTION
**Pull request by aguspina**
Fixes #579

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -31,11 +32,19 @@ def set_locale
...
   def render_not_found
-    raise ActionController::RoutingError.new('Not Found')
+    raise ActionController::RoutingError.new('Not found')
   end
 
   def from_login_callback?
-    params['controller'] == 'sessions' && params['action'] == 'callback'
+    from_sessions? && params['action'] == 'callback'
+  end
+
```
Comment (flbulgarelli) [inline app/controllers/application_controller.rb]: Perhaps we could extract those three methods to a separate concern

File: app/controllers/concerns/authorization.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 module Authorization
   def authorize!
     render file: 'layouts/login' and return if must_login
-    render_not_found if !from_login_callback?  && !can_visit?
+    raise Exceptions::OrganizationPrivateException if !from_login_callback? && !can_visit? && !from_logout?
```
Comment (flbulgarelli) [inline app/controllers/concerns/authorization.rb]: I'd try to be consistent here: we are raising 404 errors using a `render_not_found` message, but raising 403 error by explicitly throwing an exception. 

I think we should always use the same strategy, whichever it is. I think that raising the exception instead of wrapping it in a method is in the end easier since it allows you to raise it from models. 

Thus, I 'd just inline that render_not_found, and raise another custom exception in the `Exception` namespace mapped to 404, but that actually means something more specific, like `Exceptions::ContentNotAccessibleException` or something like that.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -31,11 +32,19 @@ def set_locale
 
 
   def render_not_found
-    raise ActionController::RoutingError.new('Not Found')
+    raise ActionController::RoutingError.new('Not found')
```
Comment (flbulgarelli) [inline app/controllers/application_controller.rb]: This method is not useful anymore. See my next comments. 

It also can be just replaced by `not_found`

File: app/helpers/concerns/with_links_rendering.rb [hunk 4]
```diff
@@ -9,6 +9,22 @@ def link_to_user(user)
...
+  def link_to_issues(translation)
+    "<a href='https://github.com/mumuki/mumuki-atheneum/issues/new'> #{I18n.t(translation)} </a>"
+  end
+
+  def link_to_status_codes(code)
+    "<a href='https://es.wikipedia.org/wiki/Anexo:C%C3%B3digos_de_estado_HTTP'> #{I18n.t("error_#{code}")} </a>"
+  end
+
+  def mail_to_administrator
+    mail_to Organization.current.contact_email, Organization.current.contact_email, subject: I18n.t(:permissions), body: permissions_help_email_body(current_user)
```
Comment (flbulgarelli) [inline app/helpers/concerns/with_links_rendering.rb]: :+1:

File: app/models/exam.rb [hunk 5]
```diff
@@ -30,9 +30,13 @@ def enabled_for?(user)
...
 
-  def accessible_by?(user)
-    #FIXME see with @aguspina
-    user && enabled_for?(user) && authorized?(user)
+  def access!(user)
+    if user.present?
+      raise Exceptions::ExamForbiddenException unless authorized?(user)
+      raise Exceptions::ExamGoneException unless enabled_for?(user)
+    else
+      raise Exceptions::ExamForbiddenException
```
Comment (flbulgarelli) [inline app/models/exam.rb]: If user is not present, shouldn't it be an unauthorized exception, instead of forbidden?

File: app/models/exceptions/exam_gone_exception.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+class Exceptions::ExamGoneException < StandardError
```
Comment (flbulgarelli) [inline app/models/exceptions/exam_gone_exception.rb]: I think that ruby convention is naming exceptions with `Error` suffix instead of `Exception` suffix

File: app/views/errors/not_found.html.erb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<%= content_for :breadcrumbs do %>
+    <%= home_breadcrumb %>
+<% end %>
```
Comment (flbulgarelli) [inline app/views/errors/not_found.html.erb]: I think that all the error pages have the same layout. But we are repeating it always. Can we extract a layout file here?

File: spec/models/exam_spec.rb [hunk 8]
```diff
@@ -25,14 +25,14 @@
...
+        it { expect{exam.access! user}.to raise_error(Exceptions::ExamForbiddenException) }
       end
 
       context 'authorized' do
         before { exam.authorize! user }
 
-        it { expect(exam.accessible_by? user).to be true }
-        it { expect(exam.accessible_by? other_user).to be false }
+        it { expect{exam.access! user}.to_not raise_error }
+        it { expect{exam.access! other_user}.to raise_error(Exceptions::ExamForbiddenException) }
```
Comment (flbulgarelli) [inline spec/models/exam_spec.rb]: Perhaps we could add test cases for nil user and ExamGoneException

**Comment by aguspina** [comment]
@flbulgarelli review this again please!

File: config/locales/en.yml [hunk 9]
```diff
@@ -6,13 +6,15 @@ en:
...
   author: Author
+  back_to_mumuki: Back to Mumuki!
   chapter: Chapter
   chapter_finished_html: You have finished %{chapter}!
   chapter_number: Chapter %{number} -
   clear_console: Clear console
   comment: Comment
   comments: Comments
   console: Console
+  contact_administrator: 'If you think this is not supposed to happened, please contact the admin: %{link}'
```
Comment (flbulgarelli) [inline config/locales/en.yml]: 'If you think this is not supposed to **happen**, please contact the admin'

**Comment by flbulgarelli** [comment]
:+1:
